### PR TITLE
[TAS-2422] ✨ Add search function to /store

### PR DESCRIPTION
**Pull request by AuroraHuang22**
截圖 2024-10-21 下午3 58 48
截圖 2024-10-21 下午3 58 59

### CODE REVIEW

File: src/pages/store/index.vue [hunk 1]
```diff
@@ -767,6 +931,28 @@ export default {
     selectedLanguageFilter() {
       this.scrollToTop();
     },
+    async searchQuery(newQuery) {
+      if (newQuery) {
+        try {
+          this.isSearchLoading = true;
+          this.$router.push({ query: { q: newQuery } });
```
Comment (williamchong) [inline src/pages/store/index.vue:938]:
```suggestion
          this.$router.replace({ query: { q: newQuery } });
```

File: src/pages/store/index.vue [hunk 2]
```diff
@@ -767,6 +931,28 @@ export default {
...
+          await this.fetchSearchItems(newQuery);
+        } catch (error) {
+          this.searchItems = [];
+          // eslint-disable-next-line no-console
+          console.error(error);
+        } finally {
+          this.isSearchLoading = false;
+        }
+      } else {
+        this.$router.push({ query: {} });
```
Comment (williamchong) [inline src/pages/store/index.vue:948]:
```suggestion
        this.$router.replace({ query: {} });
```

File: src/pages/store/index.vue [hunk 3]
```diff
@@ -222,6 +318,62 @@
...
+              :item-id="item.classId"
+              class-cover-frame-aspect-ratio="aspect-[4/5]"
+              :is-link-disabled="item.isMultiple"
+              @click-cover="handleClickItem($event, item)"
+            />
+          </li>
+        </ul>
+        <!-- Search not found -->
+        <div
+          v-if="searchQuery && !sortedBookstoreItems.length"
```
Comment (williamchong) [inline src/pages/store/index.vue:354]: What is `isSearchLoading`?

**Review by williamchong — CHANGES_REQUESTED** [review]

File: src/pages/store/index.vue (hunk 2 above)
Comment (AuroraHuang22) [reply src/pages/store/index.vue]: [🎨 Use router replace to store query](https://github.com/likecoin/liker-land/pull/1906/commits/840dee448f133152cbd3703f4909a3926478e56e)

File: src/locales/en.json [hunk 4]
```diff
@@ -473,6 +473,9 @@
   "listing_page_QA_content": "What is the difference between ebooks and paperbacks?",
   "listing_page_QA_label": "What is the difference between ebooks and paperbacks?",
   "listing_page_QA_title": "FAQ",
+  "listing_page_search": "Search",
+  "listing_page_search_not_found": "No related books found, please try using other keywords.",
+  "listing_page_search_recommend": "Or you can browse our recommended books",
```
[the remark this answers is not in the document]
Comment (AuroraHuang22) [reply src/locales/en.json]: [💬 Update translations](https://github.com/likecoin/liker-land/pull/1906/commits/f2da5483c8ff1fff56efd6350a8bb8113522a47d)

**Comment by AuroraHuang22** [comment]
截圖 2024-10-22 上午9 53 21

[💄 Add text for search results](https://github.com/likecoin/liker-land/pull/1906/commits/650018e47ce9b4600ac996c1ae4fef6bf687a15a)

File: src/pages/store/index.vue (hunk 3 above)
Comment (williamchong) [reply src/pages/store/index.vue:349]: @AuroraHuang22 Actually i wanted to ask if `isSearching` is set, should we hide `Search not found` message?
autocomplete issue...

Comment (AuroraHuang22) [reply src/pages/store/index.vue]: [🚸 show empty state only when search is complete](https://github.com/likecoin/liker-land/pull/1906/commits/47cfed3df4c5305071a7ec68b8d23da280702eb0)

File: src/pages/store/index.vue [hunk 5]
```diff
@@ -425,6 +576,11 @@ export default {
       isShowSortingDialog: false,
       isShowFilterDialog: false,
       dialogNFTClassList: [],
+
+      searchQuery: this.$route.query.q,
+      isSearchOpen: !!this.$route.query.q,
```
Comment (nwingt) [inline src/pages/store/index.vue:581]: These 2 should be computed props

File: src/components/Search.vue [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      >
+        <IconClose class="cursor-auto" />
+      </ButtonV2>
+    </div>
+  </div>
+</template>
+
+<script>
+import { logTrackerEvent } from '~/util/EventLogger';
+import debounce from 'lodash.debounce';
```
Comment (nwingt) [inline src/components/Search.vue:66]: absolute import first
```suggestion
import debounce from 'lodash.debounce';

import { logTrackerEvent } from '~/util/EventLogger';
```

File: src/pages/store/index.vue [hunk 7]
```diff
@@ -31,50 +31,7 @@
...
-                v-if="isSearchOpen"
-                size="tiny"
-                preset="plain"
-                @click="toggleSearch"
-              >
-                <IconClose class="cursor-auto" />
-              </ButtonV2>
-            </div>
-
+            <Search />
```
Comment (nwingt) [inline src/pages/store/index.vue:34]: Prefer 2 words for components, [ref](https://eslint.vuejs.org/rules/multi-word-component-names)
```suggestion
            <SearchBar />
```

File: src/pages/store/index.vue [hunk 8]
```diff
@@ -915,6 +835,9 @@ export default {
         content,
       }));
     },
+    searchQuery() {
+      return this.$route.query?.q || '';
+    },
```
Comment (nwingt) [inline src/pages/store/index.vue:840]: Should we remove `searchQuery` here?

File: src/pages/store/index.vue [hunk 9]
```diff
@@ -928,19 +851,7 @@ export default {
...
-        } catch (error) {
-          this.searchItems = [];
-          // eslint-disable-next-line no-console
-          console.error(error);
-        } finally {
-          this.isSearching = false;
-        }
-      } else {
-        this.$router.replace({ query: {} });
+        await this.fetchSearchItems(newQuery);
```
Comment (nwingt) [inline src/pages/store/index.vue:854]: I think we should use `@search` event in `<SearchBar />`

File: src/components/Search.vue [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      this.isSearchOpen = true;
+    }
+  },
+  methods: {
+    toggleSearch() {
+      if (this.isSearchOpen) {
+        this.$router.replace({ query: { q: undefined } });
+        logTrackerEvent(this, 'listing', 'search_clear', '', 1);
+      } else {
+        logTrackerEvent(this, 'listing', 'search_open', '', 1);
```
Comment (nwingt) [inline src/components/Search.vue:90]: I think we should emit events e.g. `@open` and `@close` and log the event from page

File: src/components/Search.vue [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+export default {
+  data() {
+    return {
+      isSearchOpen: false,
+    };
+  },
+  computed: {
+    searchQuery() {
+      return this.$route.query.q || '';
+    },
```
Comment (nwingt) [inline src/components/Search.vue:77]: What I meant is this:
```suggestion
    searchQuery: {
      get() {
        return this.$route.query.q || '';
      },
      set(value) {
       this.$router.replace({ query: { q: value } });
      },
    },
```
so you can use `this.searchQuery = 'keyword';` and no repeating `this.$router.replace`

File: src/components/Search.vue [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  },
+  mounted() {
+    if (this.searchQuery) {
+      this.isSearchOpen = true;
+    }
+  },
+  methods: {
+    toggleSearch() {
+      if (this.isSearchOpen) {
+        this.$router.replace({ query: { q: undefined } });
```
Comment (williamchong) [inline src/components/Search.vue:87]: Does query other than `q` unexpectedly get replaced here too?

File: src/components/Search.vue [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        logTrackerEvent(this, 'listing', 'search_clear', '', 1);
+      } else {
+        logTrackerEvent(this, 'listing', 'search_open', '', 1);
+      }
+      this.isSearchOpen = !this.isSearchOpen;
+    },
+    debouncedUpdateSearchKeyword: debounce(
+      function debouncedUpdateSearchKeyword(event) {
+        this.$router.replace({ query: { q: event.target.value } });
```
Comment (williamchong) [inline src/components/Search.vue:96]: Does query other than `q` unexpectedly get replaced here too?

**Review by nwingt — APPROVED** [review]

**Pull request merged**